ibis/expr/rules.py

```diff
@@ -801,3 +807,120 @@ def _validate(self, args, i):
 
 def comparable(left, right):
     return ir.castable(left, right) or ir.castable(right, left)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class TableColumnValidator():
+    @abc.abstractmethod
+    def validate(self):
+        pass
+
+
+class SubsetValidator(TableColumnValidator):
+    def __init__(self, *rules):
+        self.rules = rules
+
+    def validate(self, arg):
+        if isinstance(arg, ir.TableExpr):
+            # Check that columns match the schema first
+            for column_rule in self.rules:
+                # Members of a schema are arguments with a name
+                if not isinstance(column_rule, Argument):
```

Comment:
There's a thing that makes stuff into `Argument` instances somewhere so you shouldn't have to do this. It's very hard to get to the meat of the code here because of all the checking.

Reply:
Do you mean `_to_argument`? It seems like most argument subclasses call `_to_argument` directly, so we wouldn't even get it for free if I made `SubsetValidator` a child of `Argument`. Also, we always want to take an `Argument` here (can't think of anything we may want to upcast), so I think it is a good idea to check it explicitly.

Reply:
Although I agree it is hard to get to the meat of the code with all the validation. I will split the rule checking from the table validation.

Reply:
Ok, I made a new function to validate rules. Also it is now called in the constructor, not during validation.

Reply:
(I think this is the only outstanding-ish comment).